cpp/src/arrow/flight/sql/server_test.cc

```diff
@@ -0,0 +1,769 @@
+  const auto pk_table_name = ArrayFromJSON(utf8(), R"(["foreignTable"])");
+  const auto pk_column_name = ArrayFromJSON(utf8(), R"(["id"])");
+  const auto fk_catalog_name = ArrayFromJSON(utf8(), R"([null])");
+  const auto fk_schema_name = ArrayFromJSON(utf8(), R"([null])");
+  const auto fk_table_name = ArrayFromJSON(utf8(), R"(["intTable"])");
+  const auto fk_column_name = ArrayFromJSON(utf8(), R"(["foreignId"])");
+  const auto key_sequence = ArrayFromJSON(int32(), R"([0])");
+  const auto fk_key_name = ArrayFromJSON(utf8(), R"([null])");
+  const auto pk_key_name = ArrayFromJSON(utf8(), R"([null])");
+  const auto update_rule = ArrayFromJSON(uint8(), R"([3])");
+  const auto delete_rule = ArrayFromJSON(uint8(), R"([3])");
+
+  const std::shared_ptr<Table>& expected_table =
+      Table::Make(SqlSchema::GetCrossReferenceSchema(),
+                  {pk_catalog_name, pk_schema_name, pk_table_name, pk_column_name,
+                   fk_catalog_name, fk_schema_name, fk_table_name, fk_column_name,
+                   key_sequence, fk_key_name, pk_key_name, update_rule, delete_rule});
+  AssertTablesEqual(*expected_table, *table);
+}
+
+TEST_F(TestFlightSqlServer, TestCommandGetSqlInfo) {
+  const auto& sql_info_expected_results = sql::example::GetSqlInfoResultMap();
+  std::vector<int> sql_info_ids;
+  sql_info_ids.reserve(sql_info_expected_results.size());
+  for (const auto& sql_info_expected_result : sql_info_expected_results) {
+    sql_info_ids.push_back(sql_info_expected_result.first);
+  }
+
+  FlightCallOptions call_options;
+  ASSERT_OK_AND_ASSIGN(auto flight_info,
+                       sql_client->GetSqlInfo(call_options, sql_info_ids));
+  ASSERT_OK_AND_ASSIGN(
+      auto reader, sql_client->DoGet(call_options, flight_info->endpoints()[0].ticket));
+  std::shared_ptr<Table> results;
+  ASSERT_OK(reader->ReadAll(&results));
+  ASSERT_EQ(2, results->num_columns());
+  ASSERT_EQ(sql_info_ids.size(), results->num_rows());
+  const auto& col_name = results->column(0);
+  const auto& col_value = results->column(1);
+  for (int32_t i = 0; i < col_name->num_chunks(); i++) {
+    const auto* col_name_chunk_data =
+        col_name->chunk(i)->data()->GetValuesSafe<int32_t>(1);
+    const auto& col_value_chunk = col_value->chunk(i);
+    for (int64_t row = 0; row < col_value->length(); row++) {
+      ASSERT_OK_AND_ASSIGN(const auto& scalar, col_value_chunk->GetScalar(row));
+      const SqlInfoDenseUnionValidator validator(
+          reinterpret_cast<const DenseUnionScalar&>(*scalar));
+      const auto& expected_result =
+          sql_info_expected_results.at(col_name_chunk_data[row]);
+      arrow::util::visit(validator, expected_result);
+    }
+  }
+}
+
+TEST_F(TestFlightSqlServer, TestCommandGetSqlInfoNoInfo) {
+  FlightCallOptions call_options;
+  ASSERT_OK_AND_ASSIGN(auto flight_info, sql_client->GetSqlInfo(call_options, {999999}));
+  auto result = sql_client->DoGet(call_options, flight_info->endpoints()[0].ticket);
```

Comment:
note: you can make this `ASSERT_RAISES_WITH_MESSAGE_THAT(KeyError, ::testing::HasSubstr(...), sql_client->DoGet(...));`